docs/adrs/024-remote-module-invocation.md

```diff
@@ -0,0 +1,280 @@
+            task_wrapped_func = task.taskmethod(infer_func)
+            setattr(_RemoteModelInstance, method.signature.name, task_wrapped_func)
+    
+    # Add train method to class if one exists
+    if model_config.train_method:
+        train_func = partial(_RemoteModelInstance.remote_method_request, method=model_config.train_method)
+        setattr(_RemoteModelInstance, model_config.train_method.signature.name, train_func)
+    
+    # Return Model Instance
+    return _RemoteModelInstance(model_config.connection, model_config.model_path)    
+    
+
+class RemoteModelBase(ModuleBase):
+   def __init__(self, connection: Dict[str, Any], model_path: str):
+        ...
+   
+   def remote_method_request(self, method: RemoteMethodRpc, *args, **kwargs):
+        # Run the remote invocation using the information defined in the RemoteMethodRpc
+        if self.connection.protocol == "grpc":
+            <run remote `method` using grpc>
+        elif self.connection.protocol == "http":
+            <run remote `method` using http>
+```
+
+## RemoteModuleConfig
+
+```yaml
+class RemoteModuleConfig(ModuleConfig):
+  # Remote runtime information copied from the RemoteModelFinder config
+  connection: Dict[str, Any]
+
+  # Method information
+  # use list and tuples instead of a dictionary to avoid aconfig.Config error
+  task_methods: List[Tuple[type[TaskBase], List[RemoteMethodRpc]]]
+  train_method: RemoteMethodRpc
+  
+  # Source Module Information
+  module_id: str
+  module_name: str
+  model_path: str
+
+@dataclass
+class RemoteMethodRpc:
+    # full signature for this RPC
+    signature: CaikitMethodSignature
+    
+    # Request and response objects for this RPC
+    request_dm_name: str
+    response_dm_name: str
+    
+    # Either the function name of the GRPC Servicer or HTTP endpoint
+    rpc_name: str
+
+    # Only used for infer RPC types
+    input_streaming: bool
+    output_streaming: bool
+```
+
+The RemoteModuleConfig is a custom subclass of ModuleConfig, which contains a description of a Module's
+tasks, inference and train methods, and version information, as well as the connection information of 
+the remote runtime. The combination of the two allows the RemoteModelInitializer to construct a new 
+RemoteModule without having to import `caikit.runtime` or the source model. To help simplify the
+config definition and access, we decided to create a helper dataclass, `RemoteMethodRpc`, which contains
+information about a specific method and includes things like the CaikitMethodSignature, 
+request&response DataModel names, and the remote RPC name. The `RemoteMethodRpc` dataclass contains 
+all the runtime-specific assumptions  and is the main point where overlap happens. The 
+`RemoteModelFinder`, which constructs `RemoteMethodRpc`s, should not import `caikit.runtime`; 
+however, it will re-use a lot of code around the different naming schemes.
```

Comment:
It would be nice if that could be refactored so that these naming schemes can live in one place (and eventually be made less brittle)

Reply:
Yeah... it would be great if all of this could live in interfaces or something